src/file_chooser.rs

```diff
@@ -0,0 +1,474 @@
+        title: &str,
+        options: SaveFilesOptions,
+    ) -> PortalResponse<FileChooserResult> {
+        self.run(
+            handle,
+            app_id,
+            parent_window,
+            title,
+            FileChooserOptions::SaveFiles(options),
+        )
+        .await
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Msg {
+    DialogMessage(DialogMessage),
+    DialogResult(DialogResult),
+}
+
+#[derive(Clone)]
+pub(crate) struct Args {
+    pub handle: zvariant::ObjectPath<'static>,
+    pub app_id: String,
+    pub parent_window: String,
+    pub title: String,
+    pub options: FileChooserOptions,
+    pub tx: Sender<PortalResponse<FileChooserResult>>,
+}
+
+pub(crate) fn view(portal: &CosmicPortal, id: window::Id) -> cosmic::Element<crate::app::Msg> {
+    portal.file_chooser_dialog.as_ref().unwrap().view(id)
+}
+
+pub fn update_msg(
+    portal: &mut CosmicPortal,
+    msg: Msg,
+) -> cosmic::Command<app::Message<crate::app::Msg>> {
+    match msg {
+        Msg::DialogMessage(dialog_msg) => portal
+            .file_chooser_dialog
+            .as_mut()
+            .unwrap()
+            .update(dialog_msg),
+        Msg::DialogResult(dialog_res) => {
+            log::debug!("file chooser result {:?}", dialog_res);
+            let args = portal.file_chooser_args.take().unwrap();
+            let dialog = portal.file_chooser_dialog.take().unwrap();
+            let response = match dialog_res {
+                DialogResult::Cancel => PortalResponse::Cancelled,
+                DialogResult::Open(paths) => {
+                    let mut uris = Vec::with_capacity(paths.len());
+                    for path in paths {
+                        match url::Url::from_file_path(&path) {
+                            Ok(url) => uris.push(url.to_string()),
+                            Err(()) => {
+                                log::error!("failed to convert to URL: {:?}", path);
+                                //TODO: return error?
+                            }
+                        }
+                    }
+
+                    let dialog_choices = dialog.choices();
+                    let mut choices = Vec::with_capacity(dialog_choices.len());
+                    for choice in dialog_choices.iter() {
+                        match choice {
+                            DialogChoice::CheckBox { id, value, .. } => {
+                                choices.push((
+                                    id.clone(),
+                                    if *value { "true" } else { "false" }.to_string(),
+                                ));
+                            }
+                            DialogChoice::ComboBox {
+                                id,
+                                options,
+                                selected,
+                                ..
+                            } => {
+                                //TODO: what to fill in if selected option not set?
```

Comment:
https://github.com/flatpak/xdg-desktop-portal/blob/main/data/org.freedesktop.portal.FileChooser.xml implies that the portal can choose the initially selected value if no value is provided in the request, so maybe just picking the first option would be fine as a default.

Reply:
Fixed